cms/djangoapps/contentstore/signals/handlers.py

```diff
@@ -163,6 +168,42 @@ def listen_for_course_catalog_info_changed(sender, signal, **kwargs):
     )
 
 
+@receiver(XBLOCK_PUBLISHED)
+def listen_for_xblock_published(sender, signal, **kwargs):
+    """
+    Publish XBLOCK_PUBLISHED signals onto the event bus.
+    """
+    get_producer().send(
+        signal=XBLOCK_PUBLISHED, topic='xblock-published',
+        event_key_field='xblock_info.usage_key', event_data={'xblock_info': kwargs['xblock_info']},
+        event_metadata=kwargs['metadata'],
+    )
+
+
+@receiver(XBLOCK_DELETED)
+def listen_for_xblock_deleted(sender, signal, **kwargs):
+    """
+    Publish XBLOCK_DELETED signals onto the event bus.
+    """
+    get_producer().send(
+        signal=XBLOCK_DELETED, topic='xblock-deleted',
+        event_key_field='xblock_info.usage_key', event_data={'xblock_info': kwargs['xblock_info']},
+        event_metadata=kwargs['metadata'],
+    )
+
+
+@receiver(XBLOCK_DUPLICATED)
+def listen_for_xblock_duplicated(sender, signal, **kwargs):
+    """
+    Publish XBLOCK_DUPLICATED signals onto the event bus.
+    """
+    get_producer().send(
+        signal=XBLOCK_DUPLICATED, topic='xblock-duplicated',
+        event_key_field='xblock_info.usage_key', event_data={'xblock_info': kwargs['xblock_info']},
+        event_metadata=kwargs['metadata'],
+    )
```

Comment:
Thanks.
1. We hope to make it even simpler to produce to the event bus, but for now, the catalog updated example is a good guide.
2. We have added the ability to add a timestamp when producing the signal. That should be used in cases where the timestamp is in the db. The catalog event doesn’t do this yet, but will soon. Other existing events may want to update their calls to send event as well.
3. I imagine that these still require some infrastructure setup for the Kafka topic(s) in edx.org before this lands to not result in errors. I’m not sure what happens when the topic isn’t there.
4. We don’t yet share event types on the same topic. I’m not sure if these are meant to, so a shared consumer is picking these up in order? If so, there is a backlog event bus ticket that needs some research.
5. Out of curiosity, who are the intended consumers of the event bus events? Is there a design document around these new events?
6. I wonder how we should document events using the event bus specifically?

Thank you. Excited to see new event bus events!

Reply:
@robrap 

>We don’t yet share event types on the same topic. I’m not sure if these are meant to, so a shared consumer is picking these up in order? If so, there is a backlog event bus ticket that needs some research.

That would be nice, but the current way not a deal breaker. It will just require the consumers to run a process per signal/event.

>Out of curiosity, who are the intended consumers of the event bus events?

[Taxonomy-connector](https://github.com/openedx/taxonomy-connector/blob/8f00b139c1404695dd59ab7de27b44b410e7b730/taxonomy/signals/handlers.py#L51-L87) which is installed as part of course-discovery is the intended consumer here.

>We have added the ability to add a timestamp when producing the signal. That should be used in cases where the timestamp is in the db. The catalog event doesn’t do this yet, but will soon. Other existing events may want to update their calls to send event as well.

I am not sure if I follow, do we need to make some changes in the way we publish these events?

cc: @mariajgrimaldi

Reply:
@navinkarkera: I'm really excited that people are hopping on to the event bus. Thanks again for this work. I've got some additional thoughts and questions.

> That would be nice, but the current way not a deal breaker. It will just require the consumers to run a process per signal/event.

I updated the [event design section of the how-to](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Event-design) with a note on lifecycle events.

> [Taxonomy-connector](https://github.com/openedx/taxonomy-connector/blob/8f00b139c1404695dd59ab7de27b44b410e7b730/taxonomy/signals/handlers.py#L51-L87) which is installed as part of course-discovery is the intended consumer here.

Thank you. That's helpful, and I realize I'd love to see what ADR(s) describe the overall decisions or design for the whole flow. Are there additional links you could provide?

I also added a warning to the how-to (section linked above), as well as a small note to the [Observability section](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Observability) regarding maintainership for the event production. Since the 2U TNL team maintains cms, have they been reviewing ADRs/PRs as-needed, and are they aware that they will need to start monitoring this new event (if this is deployed to edx.org)?

Related, I added some tips to the [Publishing a signal section](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Publishing-a-signal) of the how-to regarding adding a temporary(?) rollout toggle. We can iterate on this idea as needed, but I wanted to note it.

Lastly, what teams are involved and where is this work going? Is this OpenCraft work that is initially intended for edX.org, or for your own deployment? Is OpenCraft also working with Kafka?

> I am not sure if I follow, do we need to make some changes in the way we publish these events?

Regarding my note about `time`, I tried adding a note and pointer to docs in the same [Publishing a signal section](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Publishing-a-signal). Let me know if that clears things up.

Reply:
@mariajgrimaldi: I added my question about documenting events as being event-bus events to this other PR comment: https://github.com/openedx/openedx-events/pull/80#issuecomment-1412759600

Reply:
@ormsbee:
1. [question] Maybe you could provide pointers on whether time should be sent, and if so, where to get it from?
2. [inform] We may go with multiple topics for this. I think you may have had event ordering concerns for cases like this.

Reply:
>I realize I'd love to see what ADR(s) describe the overall decisions or design for the whole flow. Are there additional links you could provide?

I'll create one describing the whole flow. 
@sameenfatima78 @mariajgrimaldi Should we add this to openedx-events or taxonomy-connector repository? 

> Since the 2U TNL team maintains cms, have they been reviewing ADRs/PRs as-needed, and are they aware that they will need to start monitoring this new event (if this is deployed to edx.org)?

This is to be deployed to edx.org. @sameenfatima78 and team from 2U (Not aware of the team name here) are aware of these events and they are also reviewing the PRs related to it.

>Related, I added some tips to the [Publishing a signal section](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Publishing-a-signal) of the how-to regarding adding a temporary(?) rollout toggle. We can iterate on this idea as needed, but I wanted to note it.

Thank you, I noticed it during testing. The producer works fine (creates the topic if it does not exist) but the consumers start logging errors if the topic does not exist.

>Lastly, what teams are involved and where is this work going? Is this OpenCraft work that is initially intended for edX.org, or for your own deployment?

Yes, this is OpenCraft work intended for edX.org.

>Is OpenCraft also working with Kafka?

If you mean using event-bus-kafka as backend, afaik we are not using events yet.

>Regarding my note about time, I tried adding a note and pointer to docs in the same [Publishing a signal section](https://openedx.atlassian.net/wiki/spaces/AC/pages/3508699151/How+to+start+using+the+Event+Bus#Publishing-a-signal). Let me know if that clears things up.

Yes, that does help. So we would need to send current time explicitly in this case as I don't think we have a timestamp of in db when xblock is published, deleted or duplicated.

Reply:
@robrap I have created a document describing the whole workflow where these events will be used in taxonomy-connector repo: https://github.com/openedx/taxonomy-connector/pull/140/files
Let me know if it helps or if I am missing something.

Reply:
As noted on that PR, this is a huge help for getting acquainted with this work. Consider adding it as a link to the various related PR descriptions, or at least this one.

Reply:
@robrap Thanks! Updated the description.

Reply:
@robrap Also added current time to events explicitly.

@mariajgrimaldi Let me know if we are missing something.

Reply:
> Lastly, what teams are involved and where is this work going? Is this OpenCraft work that is initially intended for edX.org, or for your own deployment?

@robrap This work is a blended project between OpenCraft and our team (Enterprise Markhors). 

> Since the 2U TNL team maintains cms, have they been reviewing ADRs/PRs as-needed, and are they aware that they will need to start monitoring this new event (if this is deployed to edx.org)?
@robrap Thank you for pointing that out.

Reply:
Thanks again. I'm not sure how work is being tracked, but maybe there could be a ticket (or tickets) for the following:
1. Checking in with TNL on event bus design and ownership for these 3 events.
2. Related, determining whether the 3 events will be sent to a single topic, or multiple topics, which may have long term implications (unless we do one, and then supplement with the other in the future).
3. Ensuring that the stage/production topics get created before starting to send events, as well as setting up monitoring/alerting, etc.

Other than that, this thread can be marked resolved from my end, as long as everyone else agrees. Thanks!

Reply:
@navinkarkera: exciting work! Thanks for all the info context. I'm happy with the state of the PR and all the discussion surrounding it. Do you know if there is anything else that should be addressed?

Reply:
hello @mphilbrick211: should someone from Enterprise Markhors approve this PR also?

Reply:
Thank you @robrap, I'll run these by our team.